File: freud/py/summarize.py
#!/usr/bin/env python3
"""Generate compact run summaries.

Port of freud/scripts/summarize-run.sh — reads indexed artifacts from a Freud
run directory and produces:
  artifacts/summary-compact.md
  artifacts/summary-compact.json
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

from freud.py.common import (
    extract_json_number,
    extract_json_string,
    head_lines,
    json_escape,
    rg_search,
)


def summarize(run_dir: str) -> str:
    """Generate compact summary for *run_dir*. Returns path to summary-compact.md."""
    run_path = Path(run_dir)
    if not run_path.is_dir():
        print(f"Run directory does not exist: {run_dir}", file=sys.stderr)
        sys.exit(1)

    artifact_dir = run_path / "artifacts"
    summary_json = artifact_dir / "summary.json"
    steps_file = artifact_dir / "step-index.tsv"
    trail_file = artifact_dir / "trail.jsonl"
    trail_index_file = artifact_dir / "trail-index.tsv"
    run_config_file = artifact_dir / "run-config.json"
    step_meta_dir = artifact_dir / "step-meta"
    freud_metrics_json = artifact_dir / "freud-metrics.json"
    summary_md = artifact_dir / "summary-compact.md"
    summary_compact_json = artifact_dir / "summary-compact.json"
    anomalies_md = artifact_dir / "anomalies.md"

    feature_id = extract_json_string(summary_json, "feature_id")
    run_id = extract_json_string(summary_json, "run_id")
    status = extract_json_string(summary_json, "status")
    mode = extract_json_string(summary_json, "mode")
    steps_failed = extract_json_number(summary_json, "steps_failed")
    failed_test_count = extract_json_number(summary_json, "failed_test_count")

    # First failed step from step-index.tsv
    first_failed_step = ""
    if steps_file.is_file():
        for line in steps_file.read_text(encoding="utf-8", errors="replace").splitlines()[1:]:
            cols = line.split("\t")
            if len(cols) >= 2 and cols[1] == "fail":
                first_failed_step = cols[0]
                break

    # Trail event count
    trail_events = 0
    if trail_file.is_file():
        trail_events = sum(
            1 for line in trail_file.read_text(encoding="utf-8", errors="replace").splitlines()
            if line.strip()
        )

    # First warning from top-signals.tsv
    top_warning_line = ""
    top_signals_path = artifact_dir / "top-signals.tsv"
    if top_signals_path.is_file():
        raw = rg_search("warning", str(top_signals_path), case_insensitive=True,
                         line_numbers=True)
        lines = [l for l in raw.splitlines() if l]
        if lines:
            top_warning_line = lines[0]

    # --- summary-compact.md ---
    md: list[str] = [
        "# Freud Compact Summary",
        "",
        f"- run_id: `{run_id or 'unknown'}`",
        f"- feature_id: `{feature_id or 'unknown'}`",
        f"- status: `{status or 'unknown'}`",
        f"- mode: `{mode or 'unknown'}`",
        f"- steps_failed: `{steps_failed or '0'}`",
        f"- failed_test_count: `{failed_test_count or '0'}`",
        f"- first_failed_step: `{first_failed_step or 'none'}`",
        f"- trail_events: `{trail_events}`",
        "",
        "## Quick Files",
        f"- `{artifact_dir}/summary.json`",
        f"- `{artifact_dir}/failures.json`",
        f"- `{artifact_dir}/anomalies.json`",
        f"- `{run_config_file}`",
        f"- `{freud_metrics_json}`",
        f"- `{artifact_dir}/step-index.tsv`",
        f"- `{trail_index_file}`",
        f"- `{artifact_dir}/trail.jsonl`",
        f"- `{step_meta_dir}/`",
        "",
    ]

    if steps_file.is_file():
        md.append("## Step Index Preview")
        md.append("```text")
        md.extend(head_lines(steps_file, 14))
        md.append("```")
        md.append("")

    if trail_index_file.is_file():
        md.append("## Trail Index Preview")
        md.append("```text")
        md.extend(head_lines(trail_index_file, 14))
        md.append("```")
        md.append("")

    if freud_metrics_json.is_file():
        md.append("## Freud Metrics Preview")
        md.extend(head_lines(freud_metrics_json, 40))
        md.append("")

    if top_warning_line:
        md.append("## First Warning Ref")
        md.append(f"- `{top_warning_line}`")
        md.append("")

    if anomalies_md.is_file():
        md.append("## Triage Preview")
        md.extend(head_lines(anomalies_md, 40))

    summary_md.write_text("\n".join(md) + "\n", encoding="utf-8")

    # --- summary-compact.json ---
    json_lines = [
        "{",
        f'  "run_id": "{json_escape(run_id)}",',
        f'  "feature_id": "{json_escape(feature_id)}",',
        f'  "status": "{json_escape(status)}",',
        f'  "mode": "{json_escape(mode)}",',
        f'  "steps_failed": {steps_failed or 0},',
        f'  "failed_test_count": {failed_test_count or 0},',
        f'  "first_failed_step": "{json_escape(first_failed_step)}",',
        f'  "trail_events": {trail_events}',
        "}",
    ]
    summary_compact_json.write_text("\n".join(json_lines) + "\n", encoding="utf-8")

    return str(summary_md)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Generate compact Freud run summary.",
        usage="freud/scripts/summarize-run.sh <run_dir>",
    )
    parser.add_argument("run_dir", help="Path to the Freud run directory")
    args = parser.parse_args()

    result = summarize(args.run_dir)
    print(result)


if __name__ == "__main__":
    main()

File: freud/internal/analysis/summarize.go
package analysis

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Summarize generates a compact summary for a run directory.
// Returns the path to summary-compact.md.
func Summarize(runDir string) (string, error) {
	if !DirExists(runDir) {
		return "", fmt.Errorf("run directory does not exist: %s", runDir)
	}

	artifactDir := filepath.Join(runDir, "artifacts")
	summaryJSON := filepath.Join(artifactDir, "summary.json")
	stepsFile := filepath.Join(artifactDir, "step-index.tsv")
	trailFile := filepath.Join(artifactDir, "trail.jsonl")
	trailIndexFile := filepath.Join(artifactDir, "trail-index.tsv")
	runConfigFile := filepath.Join(artifactDir, "run-config.json")
	stepMetaDir := filepath.Join(artifactDir, "step-meta")
	freudMetricsJSON := filepath.Join(artifactDir, "freud-metrics.json")
	summaryMD := filepath.Join(artifactDir, "summary-compact.md")
	summaryCompactJSON := filepath.Join(artifactDir, "summary-compact.json")
	anomaliesMD := filepath.Join(artifactDir, "anomalies.md")

	featureID := ExtractJSONString(summaryJSON, "feature_id")
	runID := ExtractJSONString(summaryJSON, "run_id")
	status := ExtractJSONString(summaryJSON, "status")
	mode := ExtractJSONString(summaryJSON, "mode")
	stepsFailed := ExtractJSONNumber(summaryJSON, "steps_failed")
	failedTestCount := ExtractJSONNumber(summaryJSON, "failed_test_count")

	// First failed step
	firstFailedStep := ""
	if FileExists(stepsFile) {
		data, _ := os.ReadFile(stepsFile)
		lines := strings.Split(string(data), "\n")
		for _, line := range lines[1:] { // skip header
			cols := strings.Split(line, "\t")
			if len(cols) >= 2 && cols[1] == "fail" {
				firstFailedStep = cols[0]
				break
			}
		}
	}

	// Trail event count
	trailEvents := 0
	if FileExists(trailFile) {
		trailEvents = CountLines(trailFile)
	}

	// First warning
	topWarningLine := ""
	topSignalsPath := filepath.Join(artifactDir, "top-signals.tsv")
	if FileExists(topSignalsPath) {
		raw := RgSearch("warning", topSignalsPath, true, true, "")
		lines := NonEmptyLines(raw)
		if len(lines) > 0 {
			topWarningLine = lines[0]
		}
	}

	// Build summary-compact.md
	or := func(s, fallback string) string {
		if s == "" {
			return fallback
		}
		return s
	}

	md := []string{
		"# Freud Compact Summary",
		"",
		fmt.Sprintf("- run_id: `%s`", or(runID, "unknown")),
		fmt.Sprintf("- feature_id: `%s`", or(featureID, "unknown")),
		fmt.Sprintf("- status: `%s`", or(status, "unknown")),
		fmt.Sprintf("- mode: `%s`", or(mode, "unknown")),
		fmt.Sprintf("- steps_failed: `%s`", or(stepsFailed, "0")),
		fmt.Sprintf("- failed_test_count: `%s`", or(failedTestCount, "0")),
		fmt.Sprintf("- first_failed_step: `%s`", or(firstFailedStep, "none")),
		fmt.Sprintf("- trail_events: `%d`", trailEvents),
		"",
		"## Quick Files",
		fmt.Sprintf("- `%s`", filepath.Join(artifactDir, "summary.json")),
		fmt.Sprintf("- `%s`", filepath.Join(artifactDir, "failures.json")),
		fmt.Sprintf("- `%s`", filepath.Join(artifactDir, "anomalies.json")),
		fmt.Sprintf("- `%s`", runConfigFile),
		fmt.Sprintf("- `%s`", freudMetricsJSON),
		fmt.Sprintf("- `%s`", filepath.Join(artifactDir, "step-index.tsv")),
		fmt.Sprintf("- `%s`", trailIndexFile),
		fmt.Sprintf("- `%s`", filepath.Join(artifactDir, "trail.jsonl")),
		fmt.Sprintf("- `%s/`", stepMetaDir),
		"",
	}

	if FileExists(stepsFile) {
		md = append(md, "## Step Index Preview", "```text")
		md = append(md, HeadLines(stepsFile, 14)...)
		md = append(md, "```", "")
	}

	if FileExists(trailIndexFile) {
		md = append(md, "## Trail Index Preview", "```text")
		md = append(md, HeadLines(trailIndexFile, 14)...)
		md = append(md, "```", "")
	}

	if FileExists(freudMetricsJSON) {
		md = append(md, "## Freud Metrics Preview")
		md = append(md, HeadLines(freudMetricsJSON, 40)...)
		md = append(md, "")
	}

	if topWarningLine != "" {
		md = append(md, "## First Warning Ref",
			fmt.Sprintf("- `%s`", topWarningLine), "")
	}

	if FileExists(anomaliesMD) {
		md = append(md, "## Triage Preview")
		md = append(md, HeadLines(anomaliesMD, 40)...)
	}

	os.WriteFile(summaryMD, []byte(strings.Join(md, "\n")+"\n"), 0o644)

	// Build summary-compact.json
	sfVal := or(stepsFailed, "0")
	ftcVal := or(failedTestCount, "0")
	jsonLines := []string{
		"{",
		fmt.Sprintf(`  "run_id": "%s",`, JSONEscape(runID)),
		fmt.Sprintf(`  "feature_id": "%s",`, JSONEscape(featureID)),
		fmt.Sprintf(`  "status": "%s",`, JSONEscape(status)),
		fmt.Sprintf(`  "mode": "%s",`, JSONEscape(mode)),
		fmt.Sprintf(`  "steps_failed": %s,`, sfVal),
		fmt.Sprintf(`  "failed_test_count": %s,`, ftcVal),
		fmt.Sprintf(`  "first_failed_step": "%s",`, JSONEscape(firstFailedStep)),
		fmt.Sprintf(`  "trail_events": %d`, trailEvents),
		"}",
	}
	os.WriteFile(summaryCompactJSON, []byte(strings.Join(jsonLines, "\n")+"\n"), 0o644)

	return summaryMD, nil
}
